clang/lib/Sema/SemaChecking.cpp

```diff
@@ -1900,13 +1934,157 @@ Sema::CheckBuiltinFunctionCall(FunctionDecl *FDecl, unsigned BuiltinID,
+    //  this and use the type of the second argument as the return type.
+    QualType BitsTy;
+    if (checkArgCount(*this, TheCall, 2) || checkCapArg(*this, TheCall, 0) ||
+        checkCapArg(*this, TheCall, 1, &BitsTy))
+      return ExprError();
+    // Restrict the second argument to __(u)intcap_t
+    if (!BitsTy->isIntCapType()) {
+      Diag(TheCall->getArg(1)->getExprLoc(), diag::err_typecheck_expect_int)
+          << BitsTy;
+      return ExprError();
+    }
+    // Result is always void* __capability
+    TheCall->setType(
+        Context.getPointerType(Context.VoidTy, ASTContext::PIK_Capability));
+    break;
+  }
+  case Builtin::BI__builtin_cheri_tag_clear:
+  case Builtin::BI__builtin_cheri_seal_entry: {
+    // Tag-clear and seal-entry behave like the mutator functions but don't have
+    QualType SrcTy;
+    // a second argument.
+    if (checkArgCount(*this, TheCall, 1) ||
+        checkCapArg(*this, TheCall, 0, &SrcTy))
+      return ExprError();
+    TheCall->setType(SrcTy);
+    break;
+  }
+  case Builtin::BI__builtin_cheri_subset_test: {
+    // For subset testing we allow any capability type for both arguments.
+    if (checkArgCount(*this, TheCall, 2) || checkCapArg(*this, TheCall, 0) ||
+        checkCapArg(*this, TheCall, 1))
+      return ExprError();
+    break;
+  }
+  case Builtin::BI__builtin_cheri_address_get:
+  case Builtin::BI__builtin_cheri_base_get:
+  case Builtin::BI__builtin_cheri_flags_get:
+  case Builtin::BI__builtin_cheri_length_get:
+  case Builtin::BI__builtin_cheri_offset_get:
+  case Builtin::BI__builtin_cheri_perms_get:
+  case Builtin::BI__builtin_cheri_sealed_get:
+  case Builtin::BI__builtin_cheri_tag_get:
+  case Builtin::BI__builtin_cheri_type_get: {
+    // The CHERI accessors should accept both capability pointer types and
+    // (u)intcap_t arguments.
+    if (checkArgCount(*this, TheCall, 1) || checkCapArg(*this, TheCall, 0))
+      return ExprError();
+    break;
+  }
+  case Builtin::BI__builtin_cheri_cap_to_pointer: {
     if (Context.getTargetInfo().areAllPointersCapabilities()) {
       Diag(TheCall->getBeginLoc(), diag::err_builtin_target_unsupported)
           << TheCall->getSourceRange();
       return ExprError();
     }
+    // First argument is the authorizing capability, the second is the
+    // capability to be converted to a relative integer pointer.
+    // For this builtin we reject __(u)intcap_t arguments as the second arg.
+    QualType SrcTy;
+    if (checkArgCount(*this, TheCall, 2) || checkCapArg(*this, TheCall, 0) ||
+        checkCapArg(*this, TheCall, 1, &SrcTy))
+      return ExprError();
+    if (!SrcTy->isPointerType())
+      return ExprError(Diag(TheCall->getBeginLoc(),
+                            diag::err_typecheck_expect_pointer_operand)
+                       << SrcTy << TheCall->getSourceRange());
+    TheCall->setType(Context.getPointerType(SrcTy->getPointeeType(),
+                                            ASTContext::PIK_Integer));
```

Comment:
Does this work in purecap? I'd expect this to crash and burn somewhere, or at least get very confused. I imagine we want to use a `uintX_t` as the result for purecap? That or we just ban the builtin and catch it earlier.

Reply:
It already returns ExprError() for purecap (line 2028).

Reply:
Ah ok, good, somehow missed that. We can ponder a `uintX_t`-returning version in purecap someday, but so long as it gives an error in a controlled manner then I'm happy.